src/lib.rs

```diff
@@ -112,42 +113,90 @@ impl ProcessManager {
 		Ok(())
 	}
 
+	/// Send a message to a process over stdin
+	pub async fn send_message(&self, key: ProcessKey, message: Cow<'static, [u8]>) -> Result<()> {
+		let inner = self.inner.read().await;
+		let process = inner.processes.get(key).ok_or(Error::NonExistantProcess)?;
+		process.process.stdin_tx.send(message).await?;
+		Ok(())
+	}
+
 	pub async fn start_process(&self, mut process: Process) -> Result<ProcessKey> {
 		if self.is_stopped() {
 			return Err(Error::Stopped);
 		}
 		let mut inner = self.inner.write().await;
+		let Some(rx) = process.stdin_rx.take() else {
+            return Err(Error::MissingStdinReceiver);
+        };
 		info!(
 			"starting process '{} {} {}'",
 			process.env_text(),
 			process.exe_text(),
 			process.args_text()
 		);
-		let cancel_token = self.cancel_token.child_token();
-		let command = Command::new(&process.executable)
-			.args(&process.args)
-			.envs(process.env.clone())
-			.stdout(Stdio::piped())
-			.stderr(Stdio::piped())
-			.stdin(Stdio::null())
-			.kill_on_drop(true)
-			.spawn()
-			.map_err(Error::Process)?;
 		let callbacks = std::mem::take(&mut process.callbacks);
+		let (callback_tx, mut callback_rx) = mpsc::unbounded_channel();
+
+		let cancel_token = self.cancel_token.child_token();
+
 		let key = inner.processes.insert(ProcessData {
 			process,
 			restarts: 0,
 			cancel_token: cancel_token.clone(),
 		});
+		let process = inner.processes.get(key).unwrap();
+
+		let mut command = Command::new(&process.process.executable);
+
+		let command = unsafe {
+			command
+				.args(&process.process.args)
+				.envs(
+					process
+						.process
+						.env
+						.iter()
+						.map(|(k, v)| (k.as_str(), v.as_str())),
+				)
+				.stdout(Stdio::piped())
+				.stderr(Stdio::piped())
+				.stdin(Stdio::piped())
+				.kill_on_drop(true)
+				.pre_exec({
+					let fd_list = if let Some(fds) = callbacks.fds.as_ref() {
+						fds()
+					} else {
+						Vec::new()
+					};
+
+					move || {
+						for fd in &fd_list {
+							util::mark_as_not_cloexec(*fd)?;
+						}
+						Ok(())
+					}
+				})
+				.spawn()
+				.map_err(Error::Process)?
+		};
+		if let Some(fds) = process.process.callbacks.fds.as_ref() {
+			let fd_list = fds();
+			for fd in fd_list {
+				if let Err(err) = util::mark_as_cloexec(fd) {
```

Comment:
If we call `mark_as_not_cloexec` only in `pre_exec`, that will only apply in the spawned subprocess, so we don't needed to mark as cloexec here.

And the `mark_as_cloexec` helper should just be unnecessary now.